### ISSUE TITLE
Add URL to HTTP Basic Auth Pop Up Window

### ISSUE THREAD

**Issue by Francewhoa** [Aaron-P/HostnameTitleBar#2]
This is a suggestion for a new feature

Hostname in Window Title 2.0 (hostname-in-window-title) is presently not able to add the URL to [HTTP Basic Auth pop up](https://en.wikipedia.org/wiki/Basic_access_authentication). With large amount of HTTP Basic Auth pop ups it's hard and slow to scroll through a long list to find the correct entry.

Steps to reproduce

1. Go to https://www.httpwatch.com/httpgallery/authentication/#showExample10
2. Click on "DISPLAY IMAGE" button
3. The HTTP Basic Auth pop up window will open. Notice that the URL was not added to the HTTP Basic Auth pop up title. This is the challenge. 
![http-basic-auth](https://user-images.githubusercontent.com/119942/40858861-93b38194-6594-11e8-9913-cc49ec3d8936.jpg)
4. The suggested new feature expected result is a URL would be added to the HTTP Basic Auth pop up title.
`Authentication Required - https://<URL.HERE>`
Instead of `Authentication Required`

[keepass-helper](https://addons.mozilla.org/en-US/firefox/addon/keepass-helper/) add-on is able to do add URL to HTTP Basic Auth pop up. **Maybe some of its [open source code](https://app.assembla.com/wiki/show/slipstream/KeePass_Helper ) could be recycled for url-in-title-keepass?**

Any volunteer for a patch? I would be happy to contribute testing and or documentation.

Using:

- Hostname in Window Title 2.0
- Firefox 60.5.0esr at 64 bit
- Debian 8 Jessie

**Comment by Francewhoa** [comment]
According to  M-Gregoire
* WebExtension API does not allow to change the title of the Basic Auth window https://github.com/M-Gregoire/Firefox-AddUrlToWindowsTitle/issues/1#issuecomment-395957565

* The Basic Auth window/login prompt/dialog is a response built into the user-agent/browser and cannot be changed by the server https://github.com/M-Gregoire/Firefox-AddUrlToWindowsTitle/issues/1#issuecomment-395967794

**Comment by Francewhoa** [comment]
By the way thanks @Aaron-P for your extension :) I love it. Today I contributed a 5 stars review at https://addons.mozilla.org/en-US/firefox/addon/hostname-in-window-title/reviews/1284035/

**Comment by Aaron-P** [comment]
The title of the basic auth modal cannot be changed by a webextensions api extension.

I have done some exploring into injecting scripts into the chrome dom using [userChrome.css](https://github.com/nuchi/firefox-quantum-userchromejs) but I don't know much about the internal structure of the chrome dom.  I have spent a few hours trying to find some way to get the context of the basic auth modal window but haven't had any luck so far.  There is very little useful documentation on this, so I'm really not sure how to go about finding it or if it's even possible.

**Comment by Aaron-P** [comment]
I have found a way to kind of do it, though not as well as I would prefer.  You need to use https://github.com/nuchi/firefox-quantum-userchromejs to inject this script through userChrome.js.

    (function () {
        var observers = {
            "common-dialog-loaded": {
                aTopic: "common-dialog-loaded",
                observe: function (aSubject, aTopic, aData) {

                    var promptType = aSubject.args.promptType;
                    if (["promptUserAndPass", "promptPassword"].indexOf(promptType) === -1)
                        return;

                    var openerUrlMatch = aSubject.args.text.match(/^https?:\/\/[^ ]+/);
                    if (openerUrlMatch.length === 0)
                        return;

                    try
                    {
                        var url = new URL(openerUrlMatch[0]);
                        aSubject.document.title += " - " + url.hostname;
                    } catch (e) {}

                    //TODO: It would be nice to get the url from the opener.gBrowser.currentTab but I'm not sure how to do that on initial load.
                    // console.log(aSubject)
                    // console.log(aSubject.opener);
                    // console.log(aSubject.opener.gBrowser);
                },
                reg: function () {
                    Services.obs.addObserver(observers[this.aTopic], this.aTopic, false);
                },
                unreg: function () {
                    Services.obs.removeObserver(observers[this.aTopic], this.aTopic);
                }
            }
        };

        observers["common-dialog-loaded"].reg();
    }());

This is the product:

![capture](https://user-images.githubusercontent.com/744119/52521905-23b56700-2c44-11e9-9ace-78cdf17deeb0.PNG)

I have no idea how fragile this is.

**Comment by Francewhoa** [comment]
Thanks for your replies & efforts Aaron-P :) The present Webextensions API seems challenging to work with. Others mentioned the same challenge. In the future, it might be easier.

Congrats on finding this [workaround](#issuecomment-462048771). I would love to contribute testing this, but I'm not a software engineer. I'm able to test Firefox extensions though. If I understand correctly the suggestion workaround, it would require users to manually customize the userChrome.css file(s).

I agree that this work around is risky to be unstable. Speaking for myself I prefer using stable Firefox extensions. Or when possible use extensions that are not dependent on external script(s) which needs to be manually updates.

**Comment by Francewhoa** [comment]
This is a suggestion about another resolution. Which is inspired from your comments above. If not already done, how about opening a ticket with Firefox? Suggesting parent & child titles? I mean for Firefox to automatically pass on the "parent" page title to it's "child" HTTP Basic Auth Pop Up Window? 

In this suggested mockup below, the number 2 shows the location of the "parent" title. The number 3 shows the location of the "child" title. Which as you know is the HTTP Basic Auth Pop Up Window.
![mockup---title---parent---child---firefox---v1](https://user-images.githubusercontent.com/119942/52596789-0c25dc00-2e06-11e9-9f72-7e9d6e83df53.jpg)

In other words, if Firefox add this suggested new feature about parent & child titles, it might be easier for you from there, to potentially customize both the parent and child titles. Without being dependent on external scripts. Also it would be much easier for users. They would simply install your extension, then optionally, using your extension Templates, customized the child title :)

**Comment by Aaron-P** [comment]
Closing this unless there are changes to the Firefox api.  I have added the sample code for the userChrome.js workaround to the repo for those interested https://github.com/Aaron-P/HostnameTitleBar/tree/master/BasicAuthDialog.

**Comment by Aaron-P** [comment]
At some point Mozilla changed the way the password dialogs work to make them in-window modals instead of system modals. These in-window modals don't have a title bar, so this workaround no longer works.

Since the modals are now in-window, keepass can read off the title of the window the modal is in, partially alleviating the original issue, however, it is important to note that it will be the title state of the window that started the navigation, not the destination of the navigation. The potential impact of this is that if you go to "yourbank.com" and click a link to "evilserver.com" which uses basic authentication, then use keepass autofill, the window title that it will see will be "yourbank.com" and you can potentially send credentials to the wrong site.

Thus, it is still important to read the password dialog box to make sure you are authenticating to the domain you want, and if you are navigating from one domain to a different domain which uses basic auth, shouldn't use keepass autofill.

**Comment by Francewhoa** [comment]
>At some point Mozilla changed the way the password dialogs work to make them in-window modals instead of system modals. These in-window modals don't have a title bar, so this workaround no longer works.

Hopefully Mozilla changed this to better match the end-users' needs. Time will tell.


>Since the modals are now in-window, keepass can read off the title of the window the modal is in, partially alleviating the original issue, however, it is important to note that it will be the title state of the window that started the navigation, not the destination of the navigation. The potential impact of this is that if you go to "yourbank.com" and click a link to "evilserver.com" which uses basic authentication, then use keepass autofill, the window title that it will see will be "yourbank.com" and you can potentially send credentials to the wrong site.
>
>Thus, it is still important to read the password dialog box to make sure you are authenticating to the domain you want, and if you are navigating from one domain to a different domain which uses basic auth, shouldn't use keepass autofill.

Thanks for flagging this risk. Good catch.

**Comment by Francewhoa** [comment]
## Suggested Collaboration Between HostnameTitleBar & KeePassXC

@Aaron-P and all interested contributors :) For those using [KeePassXC](https://keepassxc.org), schmittlauch suggested a new feature for KeePassXC. Which would automatically match the HTTP Basic Auth (HBA) in-window modal with its window title dialogue's parent window. Details https://github.com/keepassxreboot/keepassxc/issues/412

In other words, this suggested resolution could potentially allow both [HostnameTitleBar](https://github.com/Aaron-P/HostnameTitleBar) & [KeePassXC](https://keepassxc.org) addons to collaborate toward resolving this challenge. 

## Scenario
1. Using Firefox 78+, user go to this test page at https://www.httpwatch.com/httpgallery/authentication/#showExample10 
    HostnameTitleBar automatically customize this above page window title. HostnameTitleBar is already able to do this. Including automatically injecting a window title into pages that don't have one. So that all pages can both be unique & easily targeted by KeePassXC.
     The number 2 in this mockup below shows this. In this example, HostnameTitleBar injected the prefix title `https://mozilla.org`. Which as you know is a unique URL.
     ![mockup---match_hba_in-window_modal_with_parent---v1](https://user-images.githubusercontent.com/119942/101966974-b1f72d00-3bce-11eb-9fb0-afd3d14a5b71.png)
2. User click on the "**DISPLAY IMAGE**" button. Which triggers the opening of this HTTP Basic Auth (HBA) in-window modal below. 
     The number 3 in this mockup below shows this.
     ![mockup---match_hba_in-window_modal_with_parent---v1](https://user-images.githubusercontent.com/119942/101966974-b1f72d00-3bce-11eb-9fb0-afd3d14a5b71.png)
3. The challenge is that Firefox changed the way HBA work to make them in-window modals instead of system modals. Notice that this HBA in-window modals that just opened above does not have a title bar.
4. To resolve this challenge, KeePassXC would check for the parent window title. In this example, https://www.httpwatch.com/httpgallery/authentication/#showExample10
    Then KeePassXC would use this as an additional identifier. Which in turn would allow KeePassXC to automatically match each HBA in-window modal with its parent page. In other words, this suggested resolution would use the parent window info by default to distinguish between windows either without name or with the same title.

## Alternative

Alternatively, if somehow a collaboration is not realistic, both [HostnameTitleBar](https://github.com/Aaron-P/HostnameTitleBar) & [KeePassXC](https://keepassxc.org) are with open source licenses. Maybe part of their code could be use and adapted? So that both [HostnameTitleBar](https://github.com/Aaron-P/HostnameTitleBar) & [KeePassXC](https://keepassxc.org) could independently resolved this challenge with HBA in-window modals.

**Comment by Aaron-P** [comment]
Can KeePassXC discern an in-window modal from the window itself somehow? I'm not sure how they could do that unless they are reading the process memory or something like that.

I have some potential ideas that may alleviate some of this by catching authentication responses using browser.webRequest.onAuthRequired.  The issues I currently see with that idea are properly keeping track of state when tabs are switched, because you can now switch tabs while the auth dialog is open, and potential issues that may arise if some other extension cancels the request later since onAuthRequired is one of the methods that requests can be canceled in.

**Comment by Francewhoa** [comment]
>Can KeePassXC discern an in-window modal from the window itself somehow? I'm not sure how they could do that unless they are reading the process memory or something like that.

Not presently with Firefox new in-window modals. This is a suggested new feature for KeePassXC at https://github.com/keepassxreboot/keepassxc/issues/412. Thus the suggested collaboration in https://github.com/Aaron-P/HostnameTitleBar/issues/2#issuecomment-743491812 above.

KeePassXC does detect in-window modals. But is not yet able to tell the modals apart. A temporary present workaround it that KeePassXC is able to display a list of all modals. Then user have to choose one. This workaround does not scale well for users with large number of modals. For example 500 or 5000 is a long list to choose one from, LOL ;).


>I have some potential ideas that may alleviate some of this by catching authentication responses using browser.webRequest.onAuthRequired. The issues I currently see with that idea are properly keeping track of state when tabs are switched, because you can now switch tabs while the auth dialog is open, and potential issues that may arise if some other extension cancels the request later since onAuthRequired is one of the methods that requests can be canceled in.

Interesting. If your Discovery turns into a potential resolution. I would be happy to contribute testing & documentation if needed.

**Comment by Francewhoa** [comment]
## Suggested Resolution: HWT Opens The Page

@Aaron-P :) This is another suggested resolution. How about Hostname In Window Title (HWT) addon would open the page? Would this allow HWT more permissions? In turn, would it be able to inject a title into the modal?

In other words, how about the parent page would be open by HWT addon instead of be open by Firefox or by the user.

Below is the same suggestion as above. But with details & visual mockups if you're interested in those.


### Scenario

1. Using Firefox, user go to this page `about:addons`. For a one time configuration of HWT `Preferences`.  Using a new checkbox & fieldset. The number 1 in this mockup below shows those. 
       
     ![mockup---HostnameTitleBar_open_preset_window---francewhoa---ksnip---v3](https://user-images.githubusercontent.com/119942/101970708-e378f400-3be0-11eb-8c78-8cb3665abe5a.png)
    By default the checkbox would be unchecked and the fieldset would be empty. So that only interested users could activate it. To view the fieldset tin line you might need to click on this mockup to zoom in. So that you see all that the user would see. 
    Suggested text for the UI:
    * Title
      `Inject title into in-window modal on this page(s)`<sup>1</sup>
    * Help text 
      `One URL per line`
    * Foot note 
      <sup>1</sup>`This optional feature adds title to in-window modal(s). Such as but not limited to HTTP Basic Authentication (HBA) modal(s). Note that when you visit any URL(s) listed below, Hostname in Window Title will automatically reload the page. So that it has the appropriate permissions to be able to insert a title into the modal. Example of a valid page URL is: https://www.exemple.com:8080/path?search=1#hash`

2. Using this new fieldset above, the user enter one URL per line. 
    The number 1 in this mockup below shows this. With two URLs.

    ![mockup---HostnameTitleBar_open_preset_window---francewhoa---ksnip---v3---fatten---with_urls](https://user-images.githubusercontent.com/119942/101970880-f2ac7180-3be1-11eb-98b8-7022c903c917.png)

   Two URLs: 

    * https://www.httpwatch.com/httpgallery/authentication/#showExample10
    * http://www.pagetutor.com/keeper/mystash/secretstuff.html

2. Using Firefox, the user go to http://www.pagetutor.com/keeper/mystash/secretstuff.html
3. In the back-end, HWT somehow automatically detect this preset URL, intercept it, then reload this URL into the same window. In this example at http://www.pagetutor.com/keeper/mystash/secretstuff.html
    Again, this steps is into the back-end. Most users would not notice this. Maybe a short flash when the page reload.
4. Because HWT opened the page, it somehow now has both additional permissions and an history of the parent's page URL.
5. HWT inject a title into this in-window modal. Using this script at https://github.com/Aaron-P/HostnameTitleBar/issues/2#issuecomment-462048771 or any other script to your liking.
    The number 3 in this mockup show this. Except that in this example, the modal title would read 
    `[#] Authentication Required - http://www.pagetutor.com/keeper/mystash/secretstuff.html [#]`

    ![mockup---title---parent---child---firefox---v1](https://user-images.githubusercontent.com/119942/52596789-0c25dc00-2e06-11e9-9f72-7e9d6e83df53.jpg)

6. Password managers, such as KeePassXC would somehow be able to target this modal above.

7. End-users are so happy that he/she jump in joy jump and click heels together, LOL ;) Because this automation saved billions of hours ;)
     https://youtu.be/C2f8mfqO43M



### Attribution

This suggestion is inspired from maximbaz and other contributors to browserpass-extension at https://github.com/browserpass/browserpass-extension#modal-http-authentication or at https://archive.md/0G46N

**Comment by Aaron-P** [comment]
That wouldn't work.  It's not an issue of window permissions its an issue that it's not an actual system window anymore, it's something that's rendered entirely within the chrome of the browser.  Because of this it doesn't matter if we can access the modal itself, it has no 'title bar' for the system to see, and thus nothing for keepass to find when it looks at window titles.

![image](https://user-images.githubusercontent.com/744119/101975792-2960a700-3c37-11eb-81ed-e4abf6e6a4c5.png)

I'm pretty sure my idea will work, I just need to figure out how to properly keep track of state.

**Comment by Francewhoa** [comment]
>That wouldn't work. It's not an issue of window permissions its an issue that it's not an actual system window anymore, it's something that's rendered entirely within the chrome of the browser. Because of this it doesn't matter if we can access the modal itself, it has no 'title bar' for the system to see, and thus nothing for keepass to find when it looks at window titles.

Interesting. I learned something new. I understand now that the new type of window/modal is entirely within the the browser. Thanks for considering this suggestion @Aaron-P :)

>I'm pretty sure my idea will work, I just need to figure out how to properly keep track of state.

I look forward to it. As you know if your Discovery-Adventure turns into a potential resolution. I would be happy to contribute testing & documentation if needed. I'm able to test Firefox on both Debian and Windows.
